packages/ifpack2/src/Ifpack2_Details_ChebyshevKernel_def.hpp

```diff
@@ -0,0 +1,511 @@
+                                     const AMatrix& m_A_,
+                                     const XVector& m_x_,
+                                     const Scalar& beta_,
+                                     const int rows_per_team_) :
+    alpha (alpha_),
+    m_w (m_w_),
+    m_d (m_d_),
+    m_b (m_b_),
+    m_A (m_A_),
+    m_x (m_x_),
+    beta (beta_),
+    rows_per_team (rows_per_team_)
+  {
+    const size_t numRows = m_A.numRows ();
+    const size_t numCols = m_A.numCols ();
+
+    TEUCHOS_ASSERT( m_w.extent (0) == m_d.extent (0) );
+    TEUCHOS_ASSERT( m_w.extent (0) == m_b.extent (0) );
+    TEUCHOS_ASSERT( numRows == size_t (m_w.extent (0)) );
+    TEUCHOS_ASSERT( numCols <= size_t (m_x.extent (0)) );
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  void operator() (const team_member& dev) const
+  {
+    using residual_value_type = typename BVector::non_const_value_type;
+    using KAT = Kokkos::ArithTraits<residual_value_type>;
+
+    Kokkos::parallel_for
+      (Kokkos::TeamThreadRange (dev, 0, rows_per_team),
+       [&] (const LO& loop) {
+         const LO lclRow =
+           static_cast<LO> (dev.league_rank ()) * rows_per_team + loop;
+         if (lclRow >= m_A.numRows ()) {
+           return;
+         }
+         const auto A_row = m_A.rowConst(lclRow);
+         const LO row_length = static_cast<LO> (A_row.length);
+         residual_value_type A_x = KAT::zero ();
+
+         Kokkos::parallel_reduce
+           (Kokkos::ThreadVectorRange (dev, row_length),
+            [&] (const LO iEntry, residual_value_type& lsum) {
+              const auto A_val = A_row.value(iEntry);
+              lsum += A_val * m_x(A_row.colidx(iEntry));
+            }, A_x);
+
+         Kokkos::single
+           (Kokkos::PerThread(dev),
+            [&] () {
+              const auto alpha_D_res =
+                alpha * m_d(lclRow) * (m_b(lclRow) - A_x);
+              if (use_beta) {
+                m_w(lclRow) = beta * m_w(lclRow) + alpha_D_res;
+              }
+              else {
+                m_w(lclRow) = alpha_D_res;
+              }
+              m_x(lclRow) += m_w(lclRow);
+            });
+       });
+  }
+};
+
+template<class ExecutionSpace>
+int64_t
+chebyshev_kernel_vector_launch_parameters (int64_t numRows,
+                                                 int64_t nnz,
+                                                 int64_t rows_per_thread,
+                                                 int& team_size,
+                                                 int& vector_length)
+{
+  using execution_space = typename ExecutionSpace::execution_space;
+
+  int64_t rows_per_team;
+  int64_t nnz_per_row = nnz/numRows;
+
+  if (nnz_per_row < 1) {
+    nnz_per_row = 1;
+  }
+
+  if (vector_length < 1) {
+    vector_length = 1;
+    while (vector_length<32 && vector_length*6 < nnz_per_row) {
+      vector_length *= 2;
+    }
+  }
+
+  // Determine rows per thread
+  if (rows_per_thread < 1) {
+#ifdef KOKKOS_ENABLE_CUDA
```

Comment:
We have way too much `#ifdef KOKKOS_ENABLE_CUDA` in Trilinos -- not sure how to generalize that here but just something to keep in mind.